### PR TITLE
Handle output file names descripency between tsc --b and actual program emit file path calculation

### LINKED ISSUE #41801: tsc --build takes wrong assumption on project rootDir
**TypeScript Version:**  4.2.0-dev.2020120

**Search Terms:** tsc, project reference, inconsistent rootdir, builds again

**Code**

https://github.com/sguillia/ts-bug-repro-II

```ts
// No code, an empty TS file is enough to reproduce.
```

```
tsc --build --verbose
tsc --build --verbose
```

**Expected behavior:**

Project was built once.

**Actual behavior:**

Project was built twice.

**Playground Link:**

No playground link. Typescript Playground does not offer the `outDir` option.

**Related Issues:**  not found, sorry

**Additional information**

Second build logs this message:

```
[14:36:00] Project 'tsconfig.json' is out of date because output file 'dist/src/index.js' does not exist
```

`dist/src/index.js` is not supposed to exist. `dist/index.js` does.

Setting the following setting in the tsconfig.json fixes the problem:

```
"rootDir": "./src"
```

But the actual behavior should be consistent even without the `rootDir` setting

As explained [here in the docs](https://www.typescriptlang.org/docs/handbook/project-references.html#build-mode-for-typescript), `tsc --build` is supposed to consistently detect whether the project is up-to-date.

### PR DESCRIPTION
**Pull request by sheetalkamat**
1. Ensures that if there is error that all files dont belong to 'rootDir' even then 'rootDir' is used as common source directory
2. Calculation of commonSourceDirectory is done for `tsc -b` that uses `getOutputFileNames` api
Fixes  #41801 and #41780

### CODE REVIEW

**Review by andrewbranch — APPROVED** [review]

File: src/testRunner/unittests/tsbuild/helpers.ts [hunk 1]
```diff
@@ -270,11 +270,12 @@ interface Symbol {
         tick: () => void;
         baseFs: vfs.FileSystem;
         newSys: TscCompileSystem;
+        cleanBuildDescripencies: TscIncremental["cleanBuildDescripencies"];
     }
     function verifyIncrementalCorrectness(input: () => VerifyIncrementalCorrectness, index: number) {
         it(`Verify emit output file text is same when built clean for incremental scenario at:: ${index}`, () => {
             const {
-                scenario, subScenario, commandLineArgs,
+                scenario, subScenario, commandLineArgs, cleanBuildDescripencies,
```
Comment (benmccann) [inline src/testRunner/unittests/tsbuild/helpers.ts:278]: huge thanks for this fix!!

do you mean Discrepancies? it looks like that misspelling may be present several places

**Comment by benmccann** [comment]
Thank you so much for the fix!! Any idea whether it might be in 4.1.3 or not until 4.2.0?

**Comment by sheetalkamat** [comment]
We would probably want to wait for 4.2 as it is not minor fix and not patch it to 4.1

**Comment by shellscape** [comment]
@sheetalkamat I don't see this in the 4.2.0 or 4.2.1 list of issues for those milestones. Can you confirm it made it into either release?

**Comment by benmccann** [comment]
Yes, I believe this is working in TypeScript 4.2 now. It appears there was a new issue introduced directly in `rollup/plugins` and TypeScript is no longer the root cause of the failure. See https://github.com/rollup/plugins/pull/818#issuecomment-785194595 for more details

**Pull request merged**